### PR TITLE
Type validation, ability to specify local port numbers during connect, and APIs to control entity migration behaviour

### PR DESCRIPTION
**Pull request by kberg-amzn**
There are numerous fixes and improvements merged into this pull request.
- Allows a local port to be specified when opening a connection to a remote endpoint over both TCP and UDP
- Adds a wrapping serializer that can be used to validate all the types and names of objects passed through the underlying serializer (especially useful in binary serialization)
- Fixes a number of bugs, multiple outstanding but incomplete connections will no longer corrupt the udp connection set
- Replicators that end up in an intermediary state will no longer assert and crash
- Entities that end up in a damaged state and fail serialization will now be reset by the remote endpoint, causing a fresh delta from baseline
- Several data structures that required ordering for correct functioning were incorrectly using unordered variants, these have been corrected and commented on
- The cl_clientPort cvar is now hooked up and utilized, allowing for operation in high security environments where certain ports may be firewalled
- Adds an interface to prevent entities from migrating between authoritative hosts where such migration would be undesirable or cause incorrect behaviour

### CODE REVIEW

File: Code/Framework/AzNetworking/AzNetworking/Framework/NetworkingSystemComponent.cpp [hunk 1]
```diff
@@ -17,6 +17,8 @@
 
 namespace AzNetworking
 {
+    AZ_CVAR(bool, net_validateSerializedTypes, true, nullptr, AZ::ConsoleFunctorFlags::Null, "Validate that all serialized types are correct");
```
Comment (AMZN-puvvadar) [inline Code/Framework/AzNetworking/AzNetworking/Framework/NetworkingSystemComponent.cpp:20]: Should this be default on?

File: Code/Framework/AzNetworking/AzNetworking/TcpTransport/TcpSocket.cpp [hunk 2]
```diff
@@ -149,49 +149,66 @@ namespace AzNetworking
...
+        {
+            struct sockaddr_in dest;
+            memset(&dest, 0, sizeof(dest));
+
+            dest.sin_family = AF_INET;
+            dest.sin_addr.s_addr = address.GetAddress(ByteOrder::Network);
+            dest.sin_port = address.GetPort(ByteOrder::Network);
 
-            if (::listen(aznumeric_cast<int32_t>(m_socketFd), SOMAXCONN) != 0)
+            if (::connect(static_cast<int32_t>(m_socketFd), (struct sockaddr*)&dest, sizeof(dest)) != 0)
```
Comment (AMZN-puvvadar) [inline Code/Framework/AzNetworking/AzNetworking/TcpTransport/TcpSocket.cpp:201]: Looks there was a missed aznumeric_cast here

File: Code/Framework/AzNetworking/AzNetworking/UdpTransport/UdpNetworkInterface.cpp [hunk 3]
```diff
@@ -124,8 +124,15 @@ namespace AzNetworking
         }
     }
 
-    ConnectionId UdpNetworkInterface::Connect(const IpAddress& remoteAddress)
+    ConnectionId UdpNetworkInterface::Connect(const IpAddress& remoteAddress, uint16_t localPort)
     {
+        if (m_connectionSet.GetConnection(remoteAddress) != nullptr)
+        {
+            AZLOG_INFO("Attempting to connect to an endpoint (%s) that already has a connection", remoteAddress.GetString().c_str());
```
Comment (AMZN-puvvadar) [inline Code/Framework/AzNetworking/AzNetworking/UdpTransport/UdpNetworkInterface.cpp:131]: Is INFO appropriate here? IMO this is WARN worthy.

File: Gems/Multiplayer/Code/Include/Multiplayer/Components/NetBindComponent.h [hunk 4]
```diff
@@ -76,7 +76,17 @@ namespace Multiplayer
...
+
+        //! Sets whether or not a netbound entity is allowed to migrate between hosts.
+        //! Use this feature carefully, as replication is spatially based. If migration is disabled chances are you want to mark the entity as always persistent as well.
+        //! See INetworkEntityManager::MarkAlwaysRelevantToClients and INetworkEntityManager::MarkAlwaysRelevantToServers
+        //! @param value whether to enable or disable host migrations for this entity
+        void SetAllowEntityMigration(EntityMigration value);
+
+        //! Retrieves whether or not the netbound entity is allowed to migrate between hosts.
+        //! @return EntityMigration::Enabled if the entity is allowed to migrate, EntityMigration::Disabled otherwise
+        EntityMigration GetAllowEntityMigration() const;
```
Comment (AMZN-puvvadar) [inline Gems/Multiplayer/Code/Include/Multiplayer/Components/NetBindComponent.h:88]: Is this an intended part of this change? Migration doesn't seem to be explicitly mentioned in the PR.

File: Code/Framework/AzNetworking/AzNetworking/Framework/NetworkingSystemComponent.cpp (hunk 1 above)
Comment (kberg-amzn) [reply Code/Framework/AzNetworking/AzNetworking/Framework/NetworkingSystemComponent.cpp:20]: Good point, I will default this to off.

File: Code/Framework/AzNetworking/AzNetworking/UdpTransport/UdpNetworkInterface.cpp (hunk 3 above)
Comment (kberg-amzn) [reply Code/Framework/AzNetworking/AzNetworking/UdpTransport/UdpNetworkInterface.cpp:131]: I could upgrade this.. ultimately it's just a timing issue in our API, of two hosts attempt to connect to each other at roughly this situation could arise, such as in a peer to peer mesh.

File: Gems/Multiplayer/Code/Include/Multiplayer/Components/NetBindComponent.h (hunk 4 above)
Comment (kberg-amzn) [reply Gems/Multiplayer/Code/Include/Multiplayer/Components/NetBindComponent.h:88]: Ah yes, it's important.. I'll add it to the PR notes, thanks!

**Review by AMZN-puvvadar — APPROVED** [review]
Looks good, I would also update the title of the PR to mention the migration related API changes so users can identify this as a relevant PR.

File: Gems/Multiplayer/Code/Source/NetworkInput/NetworkInputArray.cpp [hunk 5]
```diff
@@ -46,48 +48,56 @@ namespace Multiplayer
...
-        {
-            return false;
-        }
-
-        // For each subsequent element
-        for (uint32_t i = 1; i < m_inputs.size(); ++i)
+        if (net_useInputDeltaSerialization)
         {
-            if (serializer.GetSerializerMode() == AzNetworking::SerializerMode::WriteToObject)
+            // Use delta-serialization to compress input RPC bandwidth usage
```
Comment (AMZN-Olex) [inline Gems/Multiplayer/Code/Source/NetworkInput/NetworkInputArray.cpp:53]: Nice!

**Review by AMZN-Olex — APPROVED** [review]

**Pull request merged**